### PR TITLE
MintPage-Fixes

### PR DESCRIPTION
**Pull request by joeizang**
Fixes required as per discussion on discord

### CODE REVIEW

File: packages/react-app/src/App.jsx [hunk 1]
```diff
@@ -13,29 +13,55 @@ import externalContracts from 'contracts/external_contracts'
...
+}
+
+const defaultState = {
+  provider: new ethers.providers.Web3Provider(window.ethereum),
+  chainId: '5',
+  contractRef: externalContracts['5'].contracts.REMIX_REWARD,
+}
+
+function appStateReducer(state, actionType) {
+  if (actionType.GOERLICHAINID) {
```
Comment (yann300) [inline packages/react-app/src/App.jsx:28]: there isn't `actionType.OPTIMISMID` here too?

File: packages/react-app/src/App.jsx [hunk 2]
```diff
@@ -44,32 +70,42 @@ function App({ mainnet, localProvider, appChainId }) {
...
+      if (holderForConnectedAddress.length > 1 && connectedAddress) {
+        setConnectedAddress(holderForConnectedAddress[0])
       }
     }
     getAddress()
-  }, [userSigner])
+  }, [appState.provider, connectedAddress])
+
+  useEffect(() => {
+    appState.provider.on('chainChanged', chainId => {
```
Comment (yann300) [inline packages/react-app/src/App.jsx:98]: here you have to dispatch the right event depending on the `chainId`

File: packages/react-app/src/App.jsx [hunk 3]
```diff
@@ -93,18 +129,56 @@ function App({ mainnet, localProvider, appChainId }) {
     }
     const provider = window.ethereum
     // @ts-ignore
-    setInjectedProvider(new ethers.providers.Web3Provider(window.ethereum))
+    // setInjectedProvider(new ethers.providers.Web3Provider(window.ethereum))
+    appDispatch({ actionType: APPSTATEACTION.OPTIMISMCHAINID })
```
Comment (yann300) [inline packages/react-app/src/App.jsx:133]: by default you dispatch `OPTIMISMID` ?

File: packages/react-app/src/App.jsx [hunk 4]
```diff
@@ -93,18 +129,56 @@ function App({ mainnet, localProvider, appChainId }) {
...
+
+    provider.on('chainChanged', chainId => {
+      console.log(`chain changed to ${chainId}! updating providers`)
+      // @ts-ignore
+      // setInjectedProvider(new ethers.providers.Web3Provider(window.ethereum))
+      appDispatch({ actionType: APPSTATEACTION.OPTIMISMCHAINID })
```
Comment (yann300) [inline packages/react-app/src/App.jsx:139]: same as above 
> here you have to dispatch the right event depending on the chainId

File: packages/react-app/src/App.jsx (hunk 1 above)
Comment (joeizang) [reply packages/react-app/src/App.jsx:28]: There is, OPTIMISMCHAINID on line 18.

File: packages/react-app/src/App.jsx (hunk 2 above)
Comment (joeizang) [reply packages/react-app/src/App.jsx:98]: You are right. This is an error on my part. Fixing right away

File: packages/react-app/src/App.jsx (hunk 3 above)
Comment (joeizang) [reply packages/react-app/src/App.jsx:133]: Default is goerli actually. But I see what you are saying here. I am refactoring the code so one of these methods is left.

File: packages/react-app/src/App.jsx [hunk 5]
```diff
@@ -10,32 +10,69 @@ import IconButton from '@mui/material/IconButton'
...
+  OPTIMISMCHAINID: '10',
+}
+
+const defaultState = {
+  provider: new ethers.providers.Web3Provider(window.ethereum),
+  chainId: '5',
+  contractRef: externalContracts['5'].contracts.REMIX_REWARD,
+}
+
+function appStateReducer(state, actionType) {
```
Comment (yann300) [inline packages/react-app/src/App.jsx:28]: if actionType has actionType.GOERLICHAINID shouldn't you do something too?

File: packages/react-app/src/App.jsx [hunk 6]
```diff
@@ -44,32 +81,76 @@ function App({ mainnet, localProvider, appChainId }) {
...
+      //   'https://mainnet.infura.io/v3/1b3241e53c8d422aab3c7c0e4101de9c',
+      // )
+      // @ts-ignore
+      setLocalProvider(local)
+      // setMainnet(mainnet)
+      setLoaded(true)
+      const provider = appState.provider
+      const net = await provider.getNetwork()
+      console.log({ provider, net })
+      if (net.chainId === APPSTATEACTION.OPTIMISMCHAINID) {
```
Comment (yann300) [inline packages/react-app/src/App.jsx:105]: should't you do something if `net.chainId === APPSTATEACTION.GOERLI` ?

File: packages/react-app/src/App.jsx [hunk 7]
```diff
@@ -113,40 +215,62 @@ function App({ mainnet, localProvider, appChainId }) {
...
-    console.log({ injectedProvider })
-    setTabValue(1)
+    // console.log({ injectedProvider })
+    setTabValue(prev => prev)
     // eslint-disable-next-line
-  }, [setInjectedProvider])
+  }, [appDispatch])
 
-  /* END - SETUP METAMASK */
+  // const loadWeb3ModalGoerli = useCallback(async () => {
```
Comment (yann300) [inline packages/react-app/src/App.jsx:223]: I think you can just remove the old code

File: packages/react-app/src/components/AllowedMintCount.jsx [hunk 8]
```diff
@@ -2,36 +2,75 @@ import Typography from '@mui/material/Typography'
...
+      try {
+        const chainId = await getCurrentChainId()
+        if (chainId === 5) {
+          // create contractRef for goerli
+          const provider = new ethers.providers.Web3Provider(window.ethereum)
+          const contractReference = externalContracts['5'].contracts.REMIX_REWARD
+          const result = await allowedMinting(contractReference, provider, contractReference.address)
+          setMintCount(result)
+        }
+        const result = await allowedMinting(contractRef, localProvider, connectedAddress)
```
Comment (yann300) [inline packages/react-app/src/components/AllowedMintCount.jsx:41]: why are you calling `allowedMinting` twice?

File: packages/react-app/src/components/BadgesPaginatedSection.jsx [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+
+  useEffect(() => {
+    if (badges.length === 0) {
+      setBadges(getPaginationData(pageSize, pageNumber))
+    }
+  }, [badges.length, getPaginationData, pageNumber, pageSize])
+
+  return (
+    <>
+      {/* <Box
```
Comment (yann300) [inline packages/react-app/src/components/BadgesPaginatedSection.jsx:50]: please remove old code

**Comment by yann300** [comment]
it is still displaying a blank screen if metamask isn't installed. it should at least display the first page because one doesn't need metamask at this point

**Comment by yann300** [comment]
When connected to Metamask.
Select an another account (from metamask), the UI doens't update with the newly selected account

File: packages/react-app/src/components/AddressedCard.jsx [hunk 10]
```diff
@@ -17,7 +17,7 @@ export default function AddressedCard({ badges }) {
       >
         {badges && badges.length > 0 ? (
           badges.map(badge => {
-            const src = 'https://remix-project.mypinata.cloud/ipfs/' + badge.decodedIpfsHash
+            const src = 'https://ipfs2.ethdevops.io/ipfs/' + badge.decodedIpfsHash
```
Comment (yann300) [inline packages/react-app/src/components/AddressedCard.jsx:20]: could you rollback this change for the moment? our ipfs is not yet ready..

**Comment by joeizang** [comment]
> 

I understand. I will jump on this right away

File: packages/react-app/src/components/AddressedCard.jsx (hunk 10 above)
Comment (joeizang) [reply packages/react-app/src/components/AddressedCard.jsx:20]: Ok. I will do that

File: packages/react-app/src/App.jsx (hunk 7 above)
Comment (joeizang) [reply packages/react-app/src/App.jsx:223]: will do

File: packages/react-app/src/App.jsx [hunk 11]
```diff
@@ -10,32 +10,75 @@ import IconButton from '@mui/material/IconButton'
...
+        provider: new ethers.providers.Web3Provider(window.ethereum),
+        chainid: '10',
+        contractRef: externalContracts['10'].contracts.REMIX_REWARD,
+      }
+      return optimism
+    case '5':
+      const goerli = {
+        provider: new ethers.providers.Web3Provider(window.ethereum),
+        chainid: '1',
+        contractRef: externalContracts['1'].contracts.REMIX_REWARD,
```
Comment (yann300) [inline packages/react-app/src/App.jsx:40]: this should be `externalContracts['5'].contracts.REMIX_REWARD,`

File: packages/react-app/src/App.jsx [hunk 12]
```diff
@@ -91,20 +170,38 @@ function App({ mainnet, localProvider, appChainId }) {
...
+        // @ts-ignore
+        appDispatch({ actionType: APPSTATEACTION.GOERLICHAINID })
+        window.location.reload()
+      }
+      if (chainId === 10 || chainId === '10') {
+        // @ts-ignore
+        appDispatch({ actionType: APPSTATEACTION.OPTIMISMCHAINID })
+        window.location.reload()
+      }
+      if (chainId !== 10 || chainId !== '10' || chainId !== 5 || chainId !== '5') {
```
Comment (yann300) [inline packages/react-app/src/App.jsx:202]: there are some duplicate

File: packages/react-app/src/App.jsx [hunk 13]
```diff
@@ -113,67 +210,58 @@ function App({ mainnet, localProvider, appChainId }) {
       logoutOfWeb3Modal()
     })
 
-    console.log({ injectedProvider })
-    setTabValue(1)
+    // console.log({ injectedProvider })
+    setTabValue(prev => prev)
```
Comment (yann300) [inline packages/react-app/src/App.jsx:214]: is it being used?

**Comment by yann300** [comment]
the page still crashes if metamask is not present.
the badges should still display even if metamask isn't installed. the mint page need metamask though.

**Comment by yann300** [comment]
I can still see these 2 issues:
> it is still displaying a blank screen if metamask isn't installed. it should at least display the first page because one doesn't need metamask at this point

> When connected to Metamask.
Select an another account (from metamask), the UI doens't update with the newly selected account

File: packages/react-app/src/App.jsx (hunk 12 above)
Comment (joeizang) [reply packages/react-app/src/App.jsx]: I have since removed these lines of code. Thanks

File: packages/react-app/src/App.jsx (hunk 11 above)
Comment (joeizang) [reply packages/react-app/src/App.jsx]: I have fixed this mismatch also

File: packages/react-app/src/App.jsx (hunk 13 above)
Comment (joeizang) [reply packages/react-app/src/App.jsx:156]: This is no longer in use and I have removed it. Thanks

**Comment by yann300** [comment]
The mint page crashes is there is no metamask.

**Comment by yann300** [comment]
The address in the header should probably use a font which matches the design.

File: packages/react-app/src/components/Account.jsx [hunk 14]
```diff
@@ -181,20 +185,19 @@ export default function Account({ minimized }) {
 
   useEffect(() => {
     if (window.ethereum !== undefined) {
-      window.ethereum.on('connect', async connectInfo => {
+      window.ethereum.on('connect', connectInfo => {
         if (window.ethereum.isConnected()) {
-          setInjectedProvider(new ethers.providers.Web3Provider(window.ethereum))
+          // setInjectedProvider(new ethers.providers.Web3Provider(window.ethereum))
```
Comment (yann300) [inline packages/react-app/src/components/Account.jsx:190]: is this going to be removed?

File: packages/react-app/src/components/AllowedMintCount.jsx [hunk 15]
```diff
@@ -58,6 +58,20 @@ export default function AllowedMintCount() {
     }
   }, [allowedMinting, connectedAddress, localProvider])
 
+  useEffect(() => {
+    if (window.ethereum === undefined) return
+    const contractReference = externalContracts['5'].contracts.REMIX_REWARD
```
Comment (yann300) [inline packages/react-app/src/components/AllowedMintCount.jsx:63]: you need to handle the case where the webapp is using the optimism network

Comment (joeizang) [reply packages/react-app/src/components/AllowedMintCount.jsx:63]: apologies. This is now done.

File: packages/react-app/src/components/Account.jsx (hunk 14 above)
Comment (joeizang) [reply packages/react-app/src/components/Account.jsx:190]: Oh yes, part of clean up. Thanks

File: packages/react-app/src/App.jsx [hunk 16]
```diff
@@ -75,13 +74,13 @@ function App() {
...
   }, [userSigner])
 
-  useEffect(() => {
-    (async () => {
-      const chainInfo = await getCurrentChainId()
-      if (selectedChainId === chainInfo[0].chainid) return
-      setSelectedChainId(chainInfo[0].chainid)
-    })()
-  }, [selectedChainId])
+  // useEffect(() => {
```
Comment (yann300) [inline packages/react-app/src/App.jsx:77]: could you remove what is not needed instead of commenting out?

**Pull request merged**